Make the anti-replay feature optional

mattcaswell opened the pull request:
This fixes the issues raised in #6389.

This also adds the ability for an application add callbacks so that it can make its own decisions about whether early data will be accepted or not if it wants to manage the replay risks itself.

Technically this PR adds features related to TLSv1.3 so it will require 2 OMC approvals.

kaduk commented on doc/man3/SSL_CTX_set_options.pod:
```diff
@@ -200,6 +200,16 @@ this option is set or not CCS messages received from the peer will always be
+OpenSSL will switch on replay protection. See L<SSL_read_early_data(3)> for a
+description of the replay protection feature. Some applications may be able to
+mitigate the replay risks in other ways and therefore the built in OpenSSL
+functionality is not required. Those applications can turn this feature off by
```
I'd rather see something like "and in such cases the built-in OpenSSL functionality is not required", and maybe even a general statement noting that "anti-replay measures are required in order to comply with the TLS 1.3 specification; OpenSSl supplies an implementation that is used by default but allows the application to override the default and provide an application-specific anti-replay mechanism" to try to drive home that this is not optional and should not just be disabled without some replacement mechanism.

mattcaswell replied:
Ok, I used your text, and added the more general statement too.

kaduk commented on doc/man3/SSL_read_early_data.pod:
```diff
@@ -186,6 +199,19 @@ In the event that the current maximum early data setting for the server is
 
+Some server applications may wish to have more control over whether early data
+is accepted or not. For example to mitigate replay risks (see L</REPLAY PROTECTION>
+below). The functions SSL_CTX_set_allow_early_data_cb and
```
Maybe also say something about declining early data when the server is heavily loaded (this scenario comes up in draft-ietf-httpbis-replay).

mattcaswell replied:
Added

kaduk commented on doc/man3/SSL_read_early_data.pod:
```diff
@@ -186,6 +199,19 @@ In the event that the current maximum early data setting for the server is
+provided when the callback was first set. Returning 1 from the callback will
+allow early data and returning 0 will reject it. Note that the
+server may reject early data for other reasons in which case this callback will
+not get called. Notably the built in replay protection feature will still be
```
grammar nit: is "built in" with a space really correct here (as opposed to "built-in" or "builtin")?

mattcaswell replied:
I changed it to "built-in".

mattcaswell commented:
> I guess I'm still not 100% convinced that we want a separate SSL_OP flag as well as the callback. From a pure TLS spec-compliance perspective, you need some mechanism, and if OpenSSL isn't providing one the application must do so. Yes, there may be cases where the entire application protocol is replay-safe, but how common are those, in comparison with the "burden" of setting a callback that always returns 1?

I suspect it won't be uncommon for an entire application protocol to be replay safe. I'm also thinking about web-apps: the replay safety may be up to what the web app does rather than what happens at the HTTP layer. Therefore having the ability to switch this on or off on a config file seems desirable (which reminds me - I forgot to add the code to do that: now done). 

I also think the presence of a callback would not necessarily mean that an application wants to disable replay protection (the callback could be used for other purposes). I'd rather an explicit option to switch that protection off, rather than it happening by accident because someone implemented a callback for some other purpose. E.g. consider the case where the reason for implementing the callback is to refuse early data when the server is overloaded; without having a separate option the application would have to re-implement the standard replay protection inside its callback, even though it actually was implemented for reasons nothing to do with replay.

mattcaswell commented:
Updated commits pushed addressing feedback so far.

mattcaswell commented:
@kaduk - any more comments on this one?

Ping @openssl/omc - I'll need 2 OMC approvals for this.

vdukhovni commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -211,6 +211,17 @@ that there will be no forward secrecy for the resumed session.
+automatically detect if a session ticket has been used more than once, TLSv1.3
+has been neotiated and early data is enabled on the server. A full handshake is
+forced if a session ticket is used a second or subsequent time. This option is
+set by default and is only used by servers. Anti-replay measures are required
```
The sentence `This option is set by default...` is confusing to me. Is it relay protection or the option to turn it off that's set by default?

kaduk replied:
Yes, please clarify this.

mattcaswell replied:
Replay protection is on by default. I clarified the wording.

vdukhovni commented on doc/man1/s_server.pod:
```diff
@@ -709,6 +710,13 @@ greater than or equal to 0.
+Switches off replay protection. With replay protection on OpenSSL will
+automatically detect if a session ticket has been used more than once, TLSv1.3
+has been neotiated and early data is enabled on the server. A full handshake is
+forced if a session ticket is used a second or subsequent time.
```
Perhaps the text should explain what happens to the early data sent along with such a replay... Is it accepted? Ignored? Does the client have to re-sendt it after the full handshake?

mattcaswell replied:
I added a note about early data here.

vdukhovni commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -441,6 +452,15 @@ middleboxes that do not understand TLSv1.3 will not drop the connection. This
+has been used more than once, TLSv1.3 has been neotiated and early data is
+enabled on the server. A full handshake is forced if a session ticket is used a
+second or subsequent time. This option is set by default and is only used by
+servers. Anti-replay measures are required to comply with the TLSv1.3
```
Now I see, the previous comment resulted from cut/paste.  Why do we have both `AntiReplay` and `-no_anti_replay`?

mattcaswell replied:
SSL_CONF supports two modes of configuring stuff - command line args and configuration file commands. Almost all options appear in two forms: one for each.

vdukhovni commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -441,6 +452,15 @@ middleboxes that do not understand TLSv1.3 will not drop the connection. This
+servers. Anti-replay measures are required to comply with the TLSv1.3
+specification. Some applications may be able to mitigate the replay risks in
+other ways and in such cases the built in OpenSSL functionality is not required.
+Equivalent to B<SSL_OP_NO_ANTI_REPLAY>.
```
If it is equivalent to the `NO` option, the name is confusing.

kaduk replied:
Indeed.

mattcaswell replied:
It's actually equivalent to the inverse of the `NO` option. I clarified this.

kaduk commented on doc/man1/s_server.pod:
```diff
@@ -709,6 +710,13 @@ greater than or equal to 0.
+
+Switches off replay protection. With replay protection on OpenSSL will
+automatically detect if a session ticket has been used more than once, TLSv1.3
+has been neotiated and early data is enabled on the server. A full handshake is
```
"negotiated".  (And always feel free to toss in a serial comma!)

mattcaswell replied:
Fixed

kaduk commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -211,6 +211,17 @@ that there will be no forward secrecy for the resumed session.
+
+Switches off replay protection. With replay protection on OpenSSL will
+automatically detect if a session ticket has been used more than once, TLSv1.3
+has been neotiated and early data is enabled on the server. A full handshake is
```
(ditto)

mattcaswell replied:
Fixed.

kaduk commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -441,6 +452,15 @@ middleboxes that do not understand TLSv1.3 will not drop the connection. This
 default. Equivalent to B<SSL_OP_ENABLE_MIDDLEBOX_COMPAT>.
 
+B<AntiReplay>: If set then OpenSSL will automatically detect if a session ticket
+has been used more than once, TLSv1.3 has been neotiated and early data is
```
(typo+comma again)

levitte replied:
( I gotta say that I appreciate the consistency of the typo ;-) )

mattcaswell replied:
Fixed.

kaduk commented on doc/man3/SSL_read_early_data.pod:
```diff
@@ -186,6 +199,20 @@ In the event that the current maximum early data setting for the server is
 
+Some server applications may wish to have more control over whether early data
+is accepted or not. For example to mitigate replay risks (see L</REPLAY PROTECTION>
+below) or to decline early_data when the server is heavily loaded. The functions
```
I think this "For example..." is not a complete sentence.  Maybe join the two with ", for example,"?

mattcaswell replied:
Fixed.

kaduk commented on doc/man3/SSL_read_early_data.pod:
```diff
@@ -186,6 +199,20 @@ In the event that the current maximum early data setting for the server is
+Some server applications may wish to have more control over whether early data
+is accepted or not. For example to mitigate replay risks (see L</REPLAY PROTECTION>
+below) or to decline early_data when the server is heavily loaded. The functions
+SSL_CTX_set_allow_early_data_cb and SSL_set_allow_early_data_cb set a callback
```
Do we need ```()``` after the function names?

levitte replied:
Yes, we do

mattcaswell replied:
Fixed.

kaduk commented on ssl/ssl_conf.c:
```diff
@@ -704,6 +706,8 @@ static const ssl_switch_tbl ssl_cmd_switches[] = {
     /* no_middlebox */
     {SSL_OP_ENABLE_MIDDLEBOX_COMPAT, SSL_TFLAG_INV},
+    /* no_anti_replay */
+    {SSL_OP_NO_ANTI_REPLAY, 0}
```
We get a new trailing comma in the previous diff chunk but not one here?

mattcaswell replied:
Fixed.

mattcaswell commented:
Updates pushed addressing all comments so far.

mattcaswell commented:
Travis red cross is not relevant.

mattcaswell commented:
Any more comments @vdukhovni or @kaduk?

vdukhovni commented on doc/man1/s_server.pod:
```diff
@@ -709,6 +710,14 @@ greater than or equal to 0.
 
+=item B<-no_anti_replay>
+
+Switches off replay protection. With replay protection on OpenSSL will
```
Add a `,` after "on".

mattcaswell replied:
Done.

vdukhovni commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -211,6 +211,17 @@ that there will be no forward secrecy for the resumed session.
 
+=item B<-no_anti_replay>
+
+Switches off replay protection. With replay protection on OpenSSL will
```
Ditto.

mattcaswell replied:
Done.

vdukhovni commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -211,6 +211,17 @@ that there will be no forward secrecy for the resumed session.
+is forced if a session ticket is used a second or subsequent time. Anti-Replay
+is on by default and is only used by servers. Anti-replay measures are required
+to comply with the TLSv1.3 specification. Some applications may be able to
+mitigate the replay risks in other ways and in such cases the built in OpenSSL
```
Hyphenate "built in" to "built-in"

FdaSilvaYY replied:
I suggest 'this built-in '

mattcaswell replied:
Done (although I stuck with "the" instead of "this")

vdukhovni commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -441,6 +452,15 @@ middleboxes that do not understand TLSv1.3 will not drop the connection. This
+second or subsequent time. This option is set by default and is only used by
+servers. Anti-replay measures are required to comply with the TLSv1.3
+specification. Some applications may be able to mitigate the replay risks in
+other ways and in such cases the built in OpenSSL functionality is not required.
```
Ditto re "built-in"

mattcaswell replied:
Done

vdukhovni commented on doc/man3/SSL_CTX_set_options.pod:
```diff
@@ -200,6 +200,17 @@ this option is set or not CCS messages received from the peer will always be
+OpenSSL will switch on replay protection. See L<SSL_read_early_data(3)> for a
+description of the replay protection feature. Anti-replay measures are required
+to comply with the TLSv1.3 specification. Some applications may be able to
+mitigate the replay risks in other ways and in such cases the built in OpenSSL
```
And here.

mattcaswell replied:
Done.

vdukhovni commented on doc/man3/SSL_CTX_set_options.pod:
```diff
@@ -200,6 +200,17 @@ this option is set or not CCS messages received from the peer will always be
+to comply with the TLSv1.3 specification. Some applications may be able to
+mitigate the replay risks in other ways and in such cases the built in OpenSSL
+functionality is not required. Those applications can turn this feature off by
+setting this option. This is a server side opton only. It is ignored for
```
s/for/by/

FdaSilvaYY replied:
... server-side option...

mattcaswell replied:
Done

vdukhovni commented on doc/man3/SSL_read_early_data.pod:
```diff
@@ -186,6 +199,20 @@ In the event that the current maximum early data setting for the server is
+decision is made to accept or reject early data. The callback is provided with a
+pointer to the user data argument that was provided when the callback was first
+set. Returning 1 from the callback will allow early data and returning 0 will
+reject it. Note that the server may reject early data for other reasons in which
```
s/the server/the OpenSSL library/  (or something similar).

mattcaswell replied:
Done

FdaSilvaYY commented on apps/s_server.c:
```diff
@@ -958,6 +959,7 @@ const OPTIONS s_server_options[] = {
     {"early_data", OPT_EARLY_DATA, '-', "Attempt to read early data"},
     {"num_tickets", OPT_S_NUM_TICKETS, 'n',
      "The number of TLSv1.3 session tickets that a server will automatically  issue" },
+    {"no_anti_replay", OPT_NO_ANTI_REPLAY, '-', "Switch of anti-replay protection"},
```
Switch-off

kaduk replied:
No hyphen, though, since switch is the verb: "Switch off anti-replay protection".

mattcaswell replied:
Fixed.

vdukhovni commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -441,6 +452,15 @@ middleboxes that do not understand TLSv1.3 will not drop the connection. This
 option is set by default. A future version of OpenSSL may not set this by
 default. Equivalent to B<SSL_OP_ENABLE_MIDDLEBOX_COMPAT>.
 
+B<AntiReplay>: If set then OpenSSL will automatically detect if a session ticket
```
Here we arrive at an interface problem. The configuration file `Options` directive supports both `AntiReplay` and `-AntiReplay`, and thus, by the time the application is processing dynamic per-invocation overrides, it may no longer be true that `AntiReplay` is on by default.  Its "default" state is presumably based on whatever is found in `Options`.
If so, and to remove confusion as to why we have `AntiReplay` here, but `-no_anti_replay` in the command option syntax, we probably need both `-no_anti_replay` and `-anti_replay` in the command syntax, or else an explicit boolean value `-anti_replay=yes|no`.  This would make it possible to override whatever is in the config file.

Finally, I am not a fan of "_" in user arguments, pressing the `Shift` key is annoying.  Could this be `-[no-]anti-replay`?

mattcaswell replied:
Yes this does make sense. I suspect (although didn't check) that there are other command line args that this same argument could be applied to (but if so that would be a different PR). I added the `-anti_replay` option.

> Finally, I am not a fan of "_" in user arguments, pressing the Shift key is annoying. Could this be -[no-]anti-replay?

I have a tendency to agree in general. However in this context *all* the other command line args use "_", so I think it would be weird (and confusing) for this one to be different.

kaduk commented on doc/man3/SSL_CTX_set_options.pod:
```diff
@@ -200,6 +200,17 @@ this option is set or not CCS messages received from the peer will always be
 
+=item SSL_OP_NO_ANTI_REPLAY
+
+By default, when a server is configured for early data (i.e. max_early_data > 0),
```
Since this is a nit-fest already, comma after "i.e.".

mattcaswell replied:
Done

vdukhovni commented:
@kaduk I thought this was already answered, but it's just a matter of "SUPPORTED COMMAND LINE COMMANDS" vs. "SUPPORTED CONFIGURATION FILE COMMANDS" per the man page. So, yes, both are needed -- they are used in different places.

It seems my comment was too concise, when I said `both switches are needed` it was in the context of my other comments, that is, I believe that we need both `-no_anti_replay` **and** `-anti_replay`, because the underlying default is unknown without knowledge of what might be in the configuration file `AntiReplay` setting (if any).

mattcaswell commented:
Travis red cross is not relevant.

mattcaswell commented:
Fixup commits pushed addressing all outstanding comments.

vdukhovni commented on doc/man1/s_server.pod:
```diff
@@ -710,13 +711,13 @@ greater than or equal to 0.
-is forced if a session ticket is used a second or subsequent time. Any early
-data that was sent will be rejected.
+Switches on or off replay protection respectively. With replay protection on
+(the default), OpenSSL will automatically detect if a session ticket has been
```
Do you think it might be appropriate to instead say something along the lines of "(compiled-in default, also supports configuration file overrides)" to indicate why we need to have a switch that forces a default?  Not critical, but perhaps makes the purpose more obvious.

mattcaswell commented:
Now that #6563 is in, this will need a small tweak.

mattcaswell commented:
I rebased this to pick up the changes from #6563. I've also done the necessary tweaks to make it work with that other PR.

I've also addressed your final comment @vdukhovni. Please can you reconfirm your approval?

This still needs a second OMC approval.

vdukhovni commented on include/openssl/ssl.h:
```diff
@@ -368,6 +368,12 @@ typedef int (*SSL_verify_cb)(int preverify_ok, X509_STORE_CTX *x509_ctx);
 
+/*
+ * Switches off automatic TLSv1.3 anti-replay protection for early data. This
+ * is a server side option only (no effect on the client).
```
server-side

vdukhovni commented on doc/man1/s_server.pod:
```diff
@@ -709,6 +711,15 @@ greater than or equal to 0.
 
+=item B<-anti_replay>, B<-no_anti_replay>
+
+Switches on or off replay protection respectively. Normally, replay protection
```
s/Normally, replay/Replay/

You already say "on by default"  below, so no need for "Normally,".

vdukhovni replied:
The first sentence becomes:  "Switches replay protection on or off, respectively".

vdukhovni commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -211,6 +211,18 @@ that there will be no forward secrecy for the resumed session.
 
+=item B<-anti_replay>, B<-no_anti_replay>
+
+Switches on or off replay protection respectively. With replay protection on,
```
s/Switches on or off replay protection respectively/Switches replay protection on or off, respectively

vdukhovni commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -211,6 +211,18 @@ that there will be no forward secrecy for the resumed session.
+time. Anti-Replay is on by default unless overridden by a configuration file and
+is only used by servers. Anti-replay measures are required to comply with the
+TLSv1.3 specification. Some applications may be able to mitigate the replay
+risks in other ways and in such cases the built-in OpenSSL functionality is not
```
s/ways/ways,/

vdukhovni commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -441,6 +453,15 @@ middleboxes that do not understand TLSv1.3 will not drop the connection. This
+second or subsequent time. This option is set by default and is only used by
+servers. Anti-replay measures are required to comply with the TLSv1.3
+specification. Some applications may be able to mitigate the replay risks in
+other ways and in such cases the built-in OpenSSL functionality is not required.
```
s/ways/ways,/

vdukhovni commented on doc/man3/SSL_CTX_set_options.pod:
```diff
@@ -226,6 +226,17 @@ this option is set or not CCS messages received from the peer will always be
+
+By default, when a server is configured for early data (i.e., max_early_data > 0),
+OpenSSL will switch on replay protection. See L<SSL_read_early_data(3)> for a
+description of the replay protection feature. Anti-replay measures are required
```
s/required to comply/required for compliance/

vdukhovni commented on doc/man3/SSL_read_early_data.pod:
```diff
@@ -186,6 +199,20 @@ In the event that the current maximum early data setting for the server is
+pointer to the user data argument that was provided when the callback was first
+set. Returning 1 from the callback will allow early data and returning 0 will
+reject it. Note that the OpenSSL library may reject early data for other reasons
+in which case this callback will not get called. Notably the built-in replay
```
S/Notably/Notably,/

vdukhovni commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -211,6 +211,18 @@ that there will be no forward secrecy for the resumed session.
+once, TLSv1.3 has been negotiated, and early data is enabled on the server. A
+full handshake is forced if a session ticket is used a second or subsequent
+time. Anti-Replay is on by default unless overridden by a configuration file and
+is only used by servers. Anti-replay measures are required to comply with the
```
s/required to comply/required for compliance/

mattcaswell commented:
Pushed with nits fixed. Thanks.
